### ISSUE TITLE
max-statements does not count chained callbacks

### ISSUE THREAD

**Issue by jscharett** [eslint/eslint#7161]
**What rule do you want to change?**
max-statements

**Does this change cause the rule to produce more or fewer warnings?**
It may cause it to produce more warnings, depending on codebase.

**How will the change be implemented? (New option, new default behavior, etc.)?**
Via a new option, so as not to break current functionality.

**Please provide some example code that this change will affect:**

``` js
function test() {
    return fn1().fn2().fn3().fn4();
}
```

**What does the rule currently do for this code?**
Currently, the rule reports there are only 1 statements for the test method, so if my max is 3, it passes.

**What will the rule do after it's changed?**
With the ability to chain methods, you have the ability to combine multiple statements into 1.  While popular and valuable, it somewhat circumvents the max-statements rule.  I propose a optional change to this rule so that it has take into account chained callbacks.  So, for the above code, max-statements would report 4, which with a max of 3, would fail.

**Comment by platinumazure** [comment]
I don't agree that max-statements or max-len statements-per-line should be modified in this case. Instead, there should be a rule that detects long call chains.

**Comment by jscharett** [comment]
Well, I guess it comes down to what is considered a statement.  I'll buy into that a chained/composite statement can be considered a single statement, but isn't it ultimately a group of statements?  The fact that they are chained should not prevent counting them imo.

I think a rule to detect long chains would be a good idea,  but then it forces me to use 2 rules.  Maybe I just have the wrong idea of what a 'statement' is.

**Comment by platinumazure** [comment]
@jscharett The ECMAScript spec is very clear on what a statement is. Multiple chained calls constitute one statement. We have to be very careful about confusing our users by overloading the meaning of a term such as "statement".

Using two rules might sound painful, but really, the reason for two rules is because they would be used for two different purposes. One limits the total number of statements (as defined by ECMAScript) in a function; the other limits the number of chained method calls or property accesses in a single statement. These are clearly orthogonal concerns.

Please understand, we do want to help, but we need to do so in a way that is best for the project and best for our users.

Perhaps the rest of the team may see things differently.

@eslint/eslint-team Thoughts on how ESLint can solve the problem identified [here](https://github.com/eslint/eslint/issues/7161#issue-177211091)?

**Comment by mysticatea** [comment]
@jscharett `return fn1().fn2().fn3().fn4();` is a `ReturnStatement` which includes 4 `CallExpression`s. `CallExpression`s are not statements. Please see [the language spec](http://www.ecma-international.org/ecma-262/7.0/#prod-Statement).

I agree there are cases that long chains are bad. But I know that chained calls are a popular style. E.g., `Promise`, d3.js, jQuery, ....
So I'm not sure it's proper for core rules.

If it's in core, I imagine `max-nested-expressions` (restrict the count of nested expression nodes) or something like.

**Comment by platinumazure** [comment]
I think the team should discuss a possible new rule. The beauty about ESLint is that rules can be turned off by users if they are using jQuery or similar and don't want the rule to be used in those cases.

@jscharett If you are open to the possibility of a new rule, would you mind editing your first post to answer the questions in [this template](https://github.com/eslint/eslint/blob/master/templates/rule-proposal.md)? Thanks!

**Comment by jscharett** [comment]
Is it better to edit it or just create a new issue?

**Comment by platinumazure** [comment]
@jscharett Either way is fine by me-- maybe closing this issue and creating a new one is cleaner, but it's up to you. If you edit the first post of this issue, that's fine too. Thanks!

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
